### PR TITLE
SIL: Utilities for calculating and verifying memory lifetimes.

### PR DESCRIPTION
**Pull request by eeckstein**
In analogy to the ownership verifier, the MemoryLifetimeVerifier checks the lifetime of memory locations.
It's based on MemoryLifetime, which is a general utility for calculating the lifetime of memory locations. It can be used by optimizations, which need to compute memory lifetime.

Memory locations are limited to addresses which are guaranteed to be not aliased, like @in/inout parameters or alloc_stack.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

File: include/swift/SIL/MemoryLifetime.h [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+/// A utility for verifying memory lifetime.
+///
+/// The MemoryLifetime utility checks the lifetime of memory locations.
+/// This is limited to memory locations which are guaranteed to be not aliased,
+/// like @in or @inout parameters. Also, alloc_stack locations are handled.
+///
+/// In addition to verification, the MemoryLifetime class can be used as utility
+/// (e.g. base class) for optimizations, which need to compute memory lifetime.
+class MemoryLifetimeVerifier : public MemoryLifetime {
```
Comment (gottesmm) [inline include/swift/SIL/MemoryLifetime.h]: Does this have to be a subclass? Can we use composition instead?

File: include/swift/SIL/MemoryLifetime.h [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+  /// Debug dump a bit set .
+  static void dumpBits(const Bits &bits);
+
+};
+
+/// A utility for verifying memory lifetime.
+///
+/// The MemoryLifetime utility checks the lifetime of memory locations.
+/// This is limited to memory locations which are guaranteed to be not aliased,
+/// like @in or @inout parameters. Also, alloc_stack locations are handled.
```
Comment (gottesmm) [inline include/swift/SIL/MemoryLifetime.h]: Shouldn't this also apply to things that have an appropriate begin_access, end_access so we know that we have unique access to them?

**Review by gottesmm — COMMENTED** [review]
Some initial comments.

My two largest concerns are around not using composition and also around the memory overhead of copying/moving this struct. If it is so large and not meant to be stored in auxillary data structures, we should make it not movable and not copyable.

**Review by atrick — APPROVED** [review]
This is a great memory verifier. It's going to be essential for enabling opaque values to verify the address lowering. The code is well organized and commented and easy to read. I'm very happy with the top-level comments too. As a verification pass, it is easy understand and nicely self-contained.

My concern is how it will be used as a lifetime query utility in other passes that run in the regular -Oxxx pipeline. Almost all optimizations should operate on object lifetimes independently. There
are usually only small subset of objects that require analysis. After each analysis/transform, the pass should free the analysis state before working on an independent problem. Passes that perform transformation should not need to guess about whether the transformation has invalidated global-all-subobject memory lifetime analysis. This memory verifier could be adapted for use as a utility, but that requires work:
- lazy allocation/initialization of the block state
- limit findLocations to a single object
- convert the dataflow's round-robin iteration into path-exploration, using a
  worklist for CFG traversal.

Although, even after it's adapted to individual lifetimes, I'm not convinced that most passes should be doing full lifetime analaysis of sublocations. Passes should be able to take advantage of well-formed SIL lifetimes that aren't decomposed into subobjects, at least up to a certain point in the pipeline. Under that assumption, lifetime analysis is a trivial path-exploration, with the boundaries defined by the use-list, that doesn't require any block state at all. The problem of modifying an `@inout` enum payload without copying the payload seems like a special case that should probably be solved with a better SIL representation that doesn't keep both the enum and payload simultaneously alive... this is just a guess and I could be wrong about it.

To summarize everything above: I think the design in this PR is good as a verification pass for now, but in the comments and description I don't want to sell it to SIL developers as a general lifetime utility.

Another thing that should probably be made clear is that this verifier relies on basically well-formed lifetime properties that don't hold in general, e.g. for anything with a cast. So it isn't suitable for diagnostic passes.

File: include/swift/SIL/MemoryLifetime.h [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+/// A utility for calculating memory lifetime.
+///
+/// The MemoryLifetime utility calculates the lifetime of memory locations.
+/// It provides functions to analyze memory locations and to calculate the
+/// dataflow.
+/// Memory locations are limited to addresses which are guaranteed to be not
+/// aliased, like @in/inout parameters and alloc_stack.
+///
+/// The MemoryLifetime class can be used as base class for optimizations,
```
[the remark this answers is not in the document]
Comment (atrick) [reply include/swift/SIL/MemoryLifetime.h]: @gottesmm This actually concerns me because I don't want people reaching for it whenever they happen to work with "memory lifetimes". What did you want it for?

File: include/swift/SIL/MemoryLifetime.h (hunk 1 above)
Comment (atrick) [reply include/swift/SIL/MemoryLifetime.h]: I also meant to comment on this. Making this a subclass is deeply confusing to me. Please don't use inheritance as a mechanism for code reuse! In fact, it will be incredibly annoying that everything in the MemoryLifetime class is under protected visibility.

File: include/swift/SIL/MemoryLifetime.h (hunk 2 above)
Comment (atrick) [reply include/swift/SIL/MemoryLifetime.h]: Interesting idea. That could add some power to the verifier.

File: include/swift/SIL/MemoryLifetime.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (gottesmm) [reply include/swift/SIL/MemoryLifetime.h]: I want some sort of generally memory lifetime tool. We do this in so many places in the compiler. Why should we reimplement it so much. We need a utility for this!

[the remark this answers is not in the document]
Comment (atrick) [reply include/swift/SIL/MemoryLifetime.h]: I strongly disagree that this is the right tool for that.

File: include/swift/SIL/MemoryLifetime.h (hunk 1 above)
Comment (atrick) [reply include/swift/SIL/MemoryLifetime.h]: Let me clarify the comment on "not using inheritance". In the best design, each class stands on its own, so it's easy to adapt and compose for new situations.

First, basic principles:

A class definition aggregates state that has the same identity and lifetime. Any operation that is meaningful to use outside that class should be provided by a public interface. Operations that can be expressed solely via that interface should be usually defined outside the class.

Inheritance is useful for polymorphism (e.g. passing any kind of instruction to a routine that doesn't care about the opcode), but beyond that it's usually anti-pattern, in that it makes the code harder to understand, compose, and debug.

Here is a template that works for designing an analysis utility:
- Separate the analysis result from the transient state
- Define a short-lived analysis class manages transient state and produces the result
- Use an rvalue method ("Result && analyze() &&") that destroys the transient state when returning its result.

Example usage:
"Result result = Analysis(environment).analyze(inputs)"

e.g. see DominatedAccessAnalysis::analyze().

The way "protected" is being used in this PR is to distinguish between the operations that are valid *during* analysis vs. operations that are valid on the result of the analysis. To the extent that it is important, that should be done by separating the analysis interface from the result interface.

As a counter example, I have been attempting to debug EscapeAnalysis utility. This involves hacking in code that detects certain pattens. The fact that all the interfaces are declared protected makes it extremely cumbersome to work with for no benefit.

Another loosely related style point: I generally dislike nested class definitions. It's not too bad here because they are fairly small, but it usually does more harm than good for understanding and hacking on the code.

File: include/swift/SIL/MemoryLifetime.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (atrick) [reply include/swift/SIL/MemoryLifetime.h]: @gottesmm you need to specify what you need from a general memory lifetime tool before you can claim that this is the thing. The major source of problems in our code is not the lack of utilities but the misuse and abuse of false abstractions and people latching onto things that have a friendly name without understanding what they need. I honestly have no idea where you would want to use this, but there could be places I haven't thought of.

File: include/swift/SIL/MemoryLifetime.h (hunk 1 above)
Comment (atrick) [reply include/swift/SIL/MemoryLifetime.h]: A big problem with this approach of organizing the analysis logic into "protected" methods is that it's impossible to hack on the implementation without constantly updating the header. The header should only contain the class layout and public interface needed to access the result of the analysis. The analysis logic should be confined to the .cpp file as either "static" functions or internal classes that manage all the transient state. (This is another annoying thing about EscapeAnalysis).

**Comment by eeckstein** [comment]
@gottesmm @atrick Thanks for reviewing. I pushed a new version: I split the MemoryLifetime utility into 2 separate ones. This addresses your comment about inheritance and also makes the utilities more flexible to use.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by atrick** [comment]
@eeckstein Thanks. That reorganization makes it even more clear how well each of the three classes are nicely decoupled, and a fair amount of the declarations have moved into the .cpp file. Ultimately, I think you as the author need to be happy with it.

One more style suggestion for `class MemoryLocations`. I find it easiest to read and work with class layouts in this form:
```
class MemoryLocations
public:
  /* exposed types */
private:
  /* state */
public:
  /* ctor/dtor/copy, need to quickly eyeball that it covers state and keep it up-to-date with state */

 /* public interface, which we often refer to */

protected:
 /* internal interface that no one looks at unless they're hacking on the implementation.
}
```

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge
